File: serial_python_api/example_basic.py
"""
example_basic.py

This example demonstrates how to communicate with the GRF-250 using the serial interface using
the built-in serial service handler.

Notes:
    Requires the pySerial module.

2024 LightWare Optoelectronics (Pty) Ltd.
https://www.lightwarelidar.com
"""

import time
import grf250_serial

# --------------------------------------------------------------------------------------------------
# Main application.
# --------------------------------------------------------------------------------------------------
print("Running GRF-250 sample.")

grf250 = grf250_serial.Grf250(grf250_serial.SerialServiceHandler("COM70", 115200))

# NOTE: Only needed if running over serial interface and mode is set to 'Wait for interface'.
grf250.initiate_serial()

product_info = grf250.get_product_information()

print("Product: " + product_info.product_name)
print("Hardware: " + str(product_info.hardware_version))
print("Firmware: " + product_info.firmware_version)
print("Serial: " + product_info.serial_number)

# Setup device
grf250.set_stream(grf250_serial.StreamId.NONE)
grf250.set_update_rate(5)
distance_config = grf250_serial.DistanceConfig.ALL
grf250.set_distance_config(distance_config)

# Poll for distance data
distance_data = grf250.get_distance_data(distance_config)
print("Polled distance: " + str(distance_data.first_return_raw_mm / 1000.0) + " m")

# Stream distance data: Blocking version.
grf250.set_stream(grf250_serial.StreamId.DISTANCE)

for i in range(10):
    distance_data = grf250.wait_for_streamed_distance(distance_config, 1)

    if distance_data is not None:
        print("Streamed distance: " + str(distance_data.first_return_raw_mm / 1000.0) + " m")
    else:
        print("Stream timeout")

# Stream distance data: Non-blocking version.
for i in range(10):
    print("Attempting to get response...")

    while True:
        # NOTE: The timeout is set to 0.
        distance_data = grf250.wait_for_streamed_distance(distance_config, 0)

        if distance_data is None:
            print("Full response not received yet, waiting/doing other work...")
            # NOTE: Simulate doing some work...
            time.sleep(0.05)
        else:
            print(
                "Non-blocking streamed distance: "
                + str(distance_data.first_return_raw_mm / 1000.0)
                + " m"
            )
            break

File: serial_c_api/examples/example_basic.c
#include <stdio.h>
#include <stdlib.h>

#include "lw_serial_api_grf250.h"

#ifdef _WIN32
#include "lw_platform_win_serial.h"
#elif __linux__
#include "lw_platform_linux_serial.h"
#endif

void lw_debug_print(const char *format, ...) {
    va_list args;
    va_start(args, format);
    vprintf(format, args);
    va_end(args);
}

void check_success(lw_result result, const char *error_message) {
    if (result != LW_RESULT_SUCCESS) {
        printf("%s\n", error_message);
        exit(1);
    }
}

int main(void) {
    // ----------------------------------------------------------------------------
    // Platform related setup.
    // ----------------------------------------------------------------------------
    lw_platform_serial_device grf250;
    // Example Windows COM port: "\\\\.\\COM70"
    // Example Linux device serial port: "/dev/ttyACM0"
    check_success(lw_platform_create_serial_device("\\\\.\\COM70", 115200, &grf250), "Failed to create serial device");

    check_success(lw_grf250_initiate_serial(&grf250.device), "Failed to initiate serial\n");

    // ----------------------------------------------------------------------------
    // Get device product info.
    // ----------------------------------------------------------------------------
    lw_grf250_product_info product_info;
    check_success(lw_grf250_get_product_info(&grf250.device, &product_info), "Failed to get product info\n");

    printf("Product name: %s\n", product_info.product_name);
    printf("Hardware version: %d\n", product_info.hardware_version);
    printf("Firmware version: %d.%d.%d\n", product_info.firmware_version.major, product_info.firmware_version.minor, product_info.firmware_version.patch);
    printf("Serial number: %s\n", product_info.serial_number);

    // ----------------------------------------------------------------------------
    // Set up the device.
    // ----------------------------------------------------------------------------
    check_success(lw_grf250_set_stream(&grf250.device, LW_GRF250_STREAM_NONE), "Failed to set stream: none\n");
    check_success(lw_grf250_set_update_rate(&grf250.device, 5), "Failed to set update rate\n");

    lw_grf_distance_config distance_config = LW_GRF250_DISTANCE_CONFIG_ALL;
    check_success(lw_grf250_set_distance_config(&grf250.device, distance_config), "Failed to set distance config\n");

    // ----------------------------------------------------------------------------
    // Poll for distance data.
    // ----------------------------------------------------------------------------
    lw_grf250_distance_data distance_data;
    check_success(lw_grf250_get_distance_data(&grf250.device, &distance_data, distance_config), "Failed to get distance data\n");

    printf("Polled distance: %d mm\n", distance_data.first_return_raw_mm);

    // ----------------------------------------------------------------------------
    // Stream distance data: Blocking version.
    // ----------------------------------------------------------------------------
    check_success(lw_grf250_set_stream(&grf250.device, LW_GRF250_STREAM_DISTANCE), "Failed to set stream: distance\n");

    for (int i = 0; i < 10; ++i) {
        lw_result result = lw_grf250_wait_for_streamed_distance(&grf250.device, distance_config, &distance_data, 1000);

        if (result == LW_RESULT_SUCCESS) {
            printf("Streamed distance: %d mm\n", distance_data.first_return_raw_mm);
        } else if (result == LW_RESULT_TIMEOUT) {
            // TODO: Could be timeout or communication error.
            printf("Stream timeout\n");
        } else if (result == LW_RESULT_ERROR) {
            printf("Communication error\n");
            return 1;
        }
    }

    // ----------------------------------------------------------------------------
    // Stream distance data: Non-blocking version.
    // ----------------------------------------------------------------------------
    for (int i = 0; i < 10; ++i) {
        while (1) {
            printf("Attempting to get response...\n");
            // NOTE: The timeout is set to 0.
            lw_result result = lw_grf250_wait_for_streamed_distance(&grf250.device, distance_config, &distance_data, 0);

            if (result == LW_RESULT_SUCCESS) {
                printf("Non blocking streamed distance: %d mm\n", distance_data.first_return_raw_mm);
                break;
            } else if (result == LW_RESULT_AGAIN) {
                printf("Full response not received yet, waiting/doing other work...\n");
                lw_platform_sleep(50);
                continue;
            } else {
                printf("Communication error\n");
                return 1;
            }
        }
    }

    // ----------------------------------------------------------------------------
    // Closing down.
    // ----------------------------------------------------------------------------
    check_success(lw_grf250_set_stream(&grf250.device, LW_GRF250_STREAM_NONE), "Failed to set stream: none\n");

    printf("Sample completed\n");

    return 0;
}
